python/gink/impl/key_set.py

```diff
@@ -0,0 +1,222 @@
+        """
+        Constructor for a set proxy.
+        
+        muid: the global id of this set, created on the fly if None
+        db: database to send commits through, or last db instance created if None
+        """
+        if root:
+            muid = Muid(-1, -1, KEY_SET)
+        database = database or Database.get_last()
+        immediate = False
+        if bundler is None:
+            immediate = True
+            bundler = Bundler(comment)
+        if muid is None:
+            muid = Container._create(KEY_SET, database=database, bundler=bundler)
+        elif muid.timestamp > 0 and contents:
+            # TODO [P3] check the store to make sure that the container is defined and compatible (possibly for set as well?)
+            pass
+        Container.__init__(self, muid=muid, database=database)
+        if contents:
+            self.clear(bundler=bundler)
+            # self.update(contents, bundler=bundler)
+        if immediate and len(bundler):
+            self._database.commit(bundler)
+
+    def add(self, key: UserKey, *, bundler: Optional[Bundler]=None, comment: Optional[str]=None):
+        """ Adds a specified value to the key set """
+        return self._add_entry(key=key, value=inclusion, bundler=bundler, comment=comment)
+    
+    def update(self, keys: Iterable[UserKey], bundler: Optional[Bundler]=None, comment: Optional[str]=None):
+        """ Adds multiple specified values to the key set """
+        # Probably best to initialize a bundler here
+        for key in keys:
+            self._add_entry(key=key, value=inclusion, bundler=bundler, comment=comment)
+
+    def contains(self, key: UserKey, as_of: GenericTimestamp=None):
+        """ Returns a boolean stating whether the specified key is in the key set """
+        as_of = self._database.resolve_timestamp(as_of)
+        found = self._database.get_store().get_entry_by_key(self.get_muid(), key=key, as_of=as_of)
+
+        return False if found is None or found.builder.deletion else True
+
+    def discard(self, key: UserKey, bundler: Optional[Bundler]=None, comment: Optional[str]=None):
+        """ Deletes a specified entry from the key set """
+        return self._add_entry(key=key, value=deletion, bundler=bundler, comment=comment)
+    
+    def remove(self, key: UserKey, bundler: Optional[Bundler]=None, comment: Optional[str]=None):
+        """ Deletes a specified entry from the key set, but returns KeyError if not found """
+        as_of = self._database.get_now()
+        found = self._database.get_store().get_entry_by_key(self.get_muid(), key=key, as_of=as_of)
+        if found is None or found.builder.deletion:  # type: ignore
+            raise KeyError("Key does not exist")
+        else:
+            return self._add_entry(key=key, value=deletion, bundler=bundler, comment=comment)
+    
+    def pop(self, key: UserKey, bundler: Optional[Bundler]=None, comment: Optional[str]=None, default=None):
```

Comment:
The pop method for python sets has a different signature and semantics than the pop method for dicts; it doesn't make sense to talk about returning a corresponding value because there isn't one.  For python sets the pop method finds an key that's in the set and deletes it and returns it, otherwise raising a key error.